server/hotrod/src/main/java/org/infinispan/server/hotrod/BaseDecoder.java

```diff
@@ -62,11 +63,15 @@ public void channelActive(ChannelHandlerContext ctx) throws Exception {
    @Override
    public void channelWritabilityChanged(ChannelHandlerContext ctx) throws Exception {
       super.channelWritabilityChanged(ctx);
+      Channel channel = ctx.channel();
+      boolean writeable = channel.isWritable();
       if (trace) {
-         log.tracef("Channel %s writability changed", ctx.channel());
+         log.tracef("Channel %s writability changed to %b", channel, writeable);
+      }
+      if (writeable) {
+         server.getClientListenerRegistry().findAndWriteEvents(channel);
```

Comment:
`writeEventsIfPossible()` already checks `channel.isWritable()` before writing.

Reply:
Yes, but `findAndWriteEvents` doesn't and allocates additional objects and uses event loop resources. Is a minor optimization.

Reply:
Oops, I've seen the method name at least 10 times and only now I see that the method is also called when the channel becomes non-writable :)

But I think `findAndWriteEvents()` also needs some work, because it does the event loop thing even if there is no client sender associated with this channel, and the way we look up the client sender in the map is pretty atrocious.

Reply:
Hrmm, it only invokes `writeEventsIfPossible` if the channel that is now writeable is associated with the given event sender in `findAndWriteEvents`. And you are referring to how we have to loop through all values I am guessing :(

I didn't write this, but I don't see why we couldn't have a `ConcurrentMap<Channel, ConcurrentHashSet<BaseClientEventSender>>` or something similar.

Reply:
Ok, so I finally started reading a bit about the Netty thread model and I saw my expectations were completely wrong. I didn't realize a channel was actually bound to a single thread, I thought the event loop was only bound to a thread pool and every event arrived on another thread in the pool, something like the [5.0 thread model](https://netty.io/wiki/new-and-noteworthy-in-5.0.html#even-more-flexible-thread-model), only with regular thread pools instead of `ForkJoinPool`.